File: example/src/main/java/com/erikeldridge/tablestore/example/MainActivity.java
package com.erikeldridge.tablestore.example;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.widget.TextView;

import com.erikeldridge.tablestore.TableStore;

import java.util.Map;
import java.util.concurrent.TimeUnit;

public class MainActivity extends AppCompatActivity {
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        final TextView view = (TextView) findViewById(R.id.output);
        assert(view != null);
        final Activity activity = this;
        new Thread(new Runnable() {
            @Override
            public void run() {
                TableStore store = TableStore.open(activity);
                store.put("users/1/name", "Ms. Foo");
                store.put("users/1/phone", "+1234567890");
                store.put("users/1/email", "1@example.com", 1, TimeUnit.MINUTES);
                final Map<String, String> phoneData = store.get("users/1/phone");
                final Map<String, String> userData = store.get("users/1");
                store.close();
                activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        view.setText(activity.getString(R.string.output, phoneData,
                                userData.toString()));
                    }
                });
            }
        }).start();
    }
}

File: app/src/main/java/com/erikeldridge/queue/MainActivity.kt
package com.erikeldridge.queue

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.widget.TextView

/**
 * sizes:
 * - base: 1505867
 * - kotlin: 1505867
 * - proto: 1745382
 */
class Store(context: Context): SQLiteOpenHelper(context, "store.db", null, 1) {
    fun put(type: String, id: String, attr: String, value: String) {
        val values = ContentValues()
        values.put(columnType, type)
        values.put(columnId, id)
        values.put(columnAttr, attr)
        values.put(columnValue, value)
        writableDatabase.insertWithOnConflict(table, "null", values, SQLiteDatabase.CONFLICT_REPLACE)
    }
    fun get(type: String, id: String, attr: String): String? {
        val cursor = readableDatabase.rawQuery(
                "select $columnValue from $table where type=? and id=? and attr=?",
                arrayOf(type, id, attr))
        val values = mutableListOf<String>()
        cursor.moveToFirst()
        while (cursor.count > 0 && !cursor.isAfterLast) {
            values.add(cursor.getString(cursor.getColumnIndex(columnValue)))
            cursor.moveToNext()
        }
        cursor.close()
        return values.firstOrNull()
    }
    fun get(type: String, id: String): Map<String, String> {
        val cursor = readableDatabase.rawQuery(
                "select $columnAttr, $columnValue from $table where type=? and id=?",
                arrayOf(type, id))
        val values = mutableMapOf<String, String>()
        cursor.moveToFirst()
        while (cursor.count > 0 && !cursor.isAfterLast) {
            values.put(cursor.getString(cursor.getColumnIndex(columnAttr)),
                    cursor.getString(cursor.getColumnIndex(columnValue)))
            cursor.moveToNext()
        }
        cursor.close()
        return values
    }
    companion object {
        val table = "v1"
        val columnType = "type"
        val columnId = "id"
        val columnAttr = "attr"
        val columnValue = "value"
        val columnUpdated = "updated"
        val createTableQuery = """
        create table $table (
            $columnType string,
            $columnId string,
            $columnAttr string,
            $columnValue string,
            $columnUpdated integer(4) not null default (strftime('%s','now')),
            primary key ($columnType, $columnId, $columnAttr)
        )
        """
    }
    override fun onCreate(db: SQLiteDatabase?) {
        db?.execSQL(createTableQuery)
    }
    override fun onUpgrade(db: SQLiteDatabase?, oldVersion: Int, newVersion: Int) {
    }
}

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val phone = findViewById(R.id.phone) as TextView
        val store = Store(this)
        store.put("users", "1", "phone", "+789")
        val values = store.get("users", "1")
        val user = Models.User.newBuilder()
                .setId(1)
                .setPhone(values["phone"])
                .build()
        phone.text = user.toString()
    }
}
